include/constant_medium.hpp

```diff
@@ -31,20 +31,14 @@ class constant_medium {
     hit_material_type = phase_function;
     material_t temp_material_type;
     hit_record rec1, rec2;
-    if (!dev_visit(
-            [&](auto&& arg) {
-              return arg.hit(ctx, r, -infinity, infinity, rec1,
```

Comment:
This seems complex.
Why not just adding something like in https://github.com/keryell/muSYCL/blob/08a920319df9d656a683b1145d495056b747cb25/include/musycl/midi/controller/keylab_essential.hpp#L69 to just return false?
Actually you can go functional and write a `exec_if_not_monostate(the_previous_lambda)`

Reply:
I think having a dedicated visitor object that handles the parameter capturing avoid repeating the lambda code (we have a lambda that only calls arg.hit in multiple place). 

Actually I think the modification that would simplify the code and be more in line with the visitor pattern would be to keep the `hittable_hit_visitor` object, and move all the hit method in the corresponding overload of the `operator()` instead of delegating to the visited element `hit()` method.

Reply:
Besides, a `exec_if_not_monostate` higher order function also requires the return value in case of monostate, so instead of 

```c++
if (!dev_visit(hittable_hit_visitor(ctx, r, -infinity, infinity, rec1,
                                        temp_material_type),
                   boundary))
```

we would have 

```c++
    if (!dev_visit(
            exec_if_not_monostate(
              [&](auto&& arg) {
                return arg.hit(ctx, r, -infinity, infinity, rec1,
                               temp_material_type);
              }), false),
            boundary))
```

Which _could be_ called complex from my point of view

Reply:
I like the exec_if_not_monostate idea, because it is very generic and prevent code separation between the visit and the visitor implementation.
but with many uses of the same visitor I can get behind hittable_hit_visitor to avoid code repetition.

Reply:
> Which _could be_ called complex from my point of view

Sure, but at least it does not mean to create some full classes with constructors to store all the `hit` arguments...
Anyway, I let you come with some good designs...